### DISCUSSION TITLE
How to pass an Id to Func<ModalClosingEventArgs, Task>

### DISCUSSION THREAD

**Discussion by michaelcsikos** [ant-design-blazor/ant-design-blazor#2373] [Q&A] [answered]
The `ConfirmOptions.OnOk` requires a `Func<ModalClosingEventArgs, Task>`. How do you pass something useful to this function? Should this be an `EventCallBack` instead?

As it is, the function can't do much for me because `A field initializer cannot reference the non-static field, method, or property '_idToDelete'.`

**Comment by ElderJames** [comment]
cc @zxyao145

**Comment by zxyao145** [comment, marked as answer]
@michaelcsikos Hi, 

I'm sorry you can't pass an Id to Func<ModalClosingEventArgs, Task>, Confirm is only used to mark two values: YES or NO (OK or Cancel; true or false).

Can you describe your usage scenario in detail? There will be other solutions.

**Reply by michaelcsikos** [reply]
Hi @zxyao145. From a `Table` with a `Delete` button I want to show a [confirm delete modal](https://antblazor.com/en-US/components/modal#components-modal-demo-confirm) passing in the `context` or `context.Id` of the object to delete. Using the `await ModalService.ConfirmAsync(options)` Ant automatically shows the spinner next to `OK` when it is clicked, which is a nice user experience while we `await` the delete operation and refresh the `Table`. However, I can't see how to pass the `Id` to `Func<ModalClosingEventArgs, Task> onOk = async (e) =>`, nor can the function use a member variable, e.g. `Guid _idToDelete`.

Using the [Open Modal with async logic](https://antblazor.com/en-US/components/modal#components-modal-demo-async) I have gotten it to work, however, it is a lot more code, and I also haven't worked out how to show the `exclamation-circle` icon. Is this possible?

**Reply by zxyao145** [reply]
> Hi @zxyao145. From a `Table` with a `Delete` button I want to show a [confirm delete modal](https://antblazor.com/en-US/components/modal#components-modal-demo-confirm) passing in the `context` or `context.Id` of the object to delete. Using the `await ModalService.ConfirmAsync(options)` Ant automatically shows the spinner next to `OK` when it is clicked, which is a nice user experience while we `await` the delete operation and refresh the `Table`. However, I can't see how to pass the `Id` to `Func<ModalClosingEventArgs, Task> onOk = async (e) =>`, nor can the function use a member variable, e.g. `Guid _idToDelete`.
> 
> Using the [Open Modal with async logic](https://antblazor.com/en-US/components/modal#components-modal-demo-async) I have gotten it to work, however, it is a lot more code, and I also haven't worked out how to show the `exclamation-circle` icon. Is this possible?




I'm sorry, you must get `id` from the context, maybe you can see the demo, Mainly the `DeleteAsync` function.

```C#
@page "/"
@using System.ComponentModel
@using AntDesign.TableModels
@inject ModalService modalService

<Table TItem="Data" DataSource="@data">
    <Column @bind-Field="@context.Name">
        <a>@context.Name</a>
    </Column>
    <Column @bind-Field="@context.Age"></Column>
    <Column @bind-Field="@context.Address">
        <TitleTemplate>
            <span> <Icon Type="environment" /> Address </span>
        </TitleTemplate>
    </Column>
    <Column @bind-Field="@context.Tags">
        @foreach (var tag in context.Tags)
        {
            var color = tag.Length > 5 ? "geekblue" : "green";
            if (tag == "loser")
            {
                color = "volcano";
            }
            <Tag Color="@color">@tag</Tag>
        }
    </Column>
    <ActionColumn Title="Action">
        <Space Size=@("middle")>
            <SpaceItem>
                <a>Invite @context.Name</a>
            </SpaceItem>
            <SpaceItem>
                <Button OnClick="@(async ()=> await DeleteAsync(context.Key))">
                    Delete
                </Button>
            </SpaceItem>
        </Space>
    </ActionColumn>
</Table>

@code {
    Data[] data = new Data[]
    {
        new()
        {
            Key = "1",
            Name = "John Brown",
            Age = 32,
            Address = "New York No. 1 Lake Park",
            Tags = new[] {"nice", "developer"}
        },
        new()
        {
            Key = "2",
            Name = "Jim Green",
            Age = 42,
            Address = "London No. 1 Lake Park",
            Tags = new[] { "loser"}
        },
        new()
        {
            Key = "3",
            Name = "Joe Black",
            Age = 32,
            Address = "Sidney No. 1 Lake Park",
            Tags = new[] { "cool", "teacher" }
        }
    };
    public class Data
    {
        [DisplayName("Key")]
        public string Key { get; set; }

        [DisplayName("Name")]
        public string Name { get; set; }

        [DisplayName("Age")]
        public int Age { get; set; }

        [DisplayName("Address")]
        public string Address { get; set; }

        [DisplayName("Tags")]
        public string[] Tags { get; set; }
    }

    private RenderFragment icon = @<Icon Type="fire" Theme="outline" />
    ;

    private Task DeleteAsync(string key)
    {
        modalService.Confirm(new ConfirmOptions()
            {
                Content = "Sure?",
                Icon = icon,
                OnOk = async (arg) =>
                {
                    await Task.Delay(3000);
                    data = data.Where(e => e.Key != key).ToArray();
                    await InvokeAsync(StateHasChanged);
                }
            });

        return Task.CompletedTask;
    }
}

```

**Reply by michaelcsikos** [reply]
@zxyao145, fantastic, thanks so much.

**Reply by zxyao145** [reply]
@michaelcsikos It's my pleasure.
